c2rust-analyze/src/c_void_casts.rs

```diff
@@ -0,0 +1,378 @@
+    ///
+    /// * the lhs [`Place`] of the [`Assign`] is a [`CVoidPtr`] in the [`To`] direction,
+    ///     as in the [`To`] direction, the [`CVoidPtr`] is on the lhs.
+    ///
+    /// * the rhs [`Place`] of the [`Rvalue::Cast`] of the [`Assign`] is a [`CVoidPtr`] in the [`From`] direction,
+    ///     as in the [`From`] direction, the [`CVoidPtr`] is on the rhs.
+    ///
+    /// [`Assign`]: StatementKind::Assign
+    /// [`From`]: CVoidCastDirection::From
+    /// [`To`]: CVoidCastDirection::To
+    pub fn should_skip_stmt(&self, stmt: &Statement<'tcx>) -> bool {
+        || -> Option<bool> {
+            let (lhs, rv) = get_assign_sides(stmt)?;
+            let skip = self.to.contains(lhs) || self.from.contains(get_cast_place(rv)?);
+            Some(skip)
+        }()
+        .unwrap_or_default()
+    }
+
+    /// Insert all applicable [`*c_void`] casts
+    /// from the `mir` [`Body`] into this [`CVoidCasts`].
+    ///
+    /// That is, find all calls to:
+    ///
+    /// * `malloc`
+    /// * `calloc`
+    /// * `realloc`
+    /// * `free`
+    ///
+    /// and insert their casts to and from [`*c_void`].
+    ///
+    /// [`*c_void`]: core::ffi::c_void
+    pub fn insert_all_from_mir(&mut self, mir: &Body<'tcx>, tcx: TyCtxt<'tcx>) {
+        for bb_data in mir.basic_blocks().iter() {
+            if let Some(term) = &bb_data.terminator {
+                let term: &Terminator = term;
+                if let TerminatorKind::Call {
+                    ref func,
+                    ref args,
+                    destination,
+                    target,
+                    ..
+                } = term.kind
+                {
+                    let func_ty = func.ty(&mir.local_decls, tcx);
+
+                    // For [`CVoidCastDirection::From`], we only count
+                    // a cast from `*c_void` to an arbitrary type in the subsequent block,
+                    // searching forward.
+                    let find_first_cast_succ_block = |get_cast| {
+                        let successor_block_id = target.unwrap();
+                        mir.basic_blocks()[successor_block_id]
+                            .statements
+                            .iter()
+                            .find_map(get_cast)
+                    };
+
+                    // For [`CVoidCastDirection::From`], we only count
+                    // a cast to `*c_void` from an arbitrary type in the same block,
+                    // searching backwards.
+                    let find_last_cast_curr_block =
+                        |get_cast| bb_data.statements.iter().rev().find_map(get_cast);
+
+                    for direction in CVoidCastDirection::from_callee(ty_callee(tcx, func_ty))
+                        .iter()
+                        .copied()
+                    {
+                        use CVoidCastDirection::*;
+                        let c_void_ptr = match direction {
+                            From => destination,
```

Comment:
if this means it's casted from a void ptr, then intuitively i'd expect the destination to be the non-void place. but here the destination is the pointer. should `From` and `To` be swapped?

Reply:
I matched the previous code,

```rust
Some(Malloc | Calloc) => {
    handle_c_void_cast(&destination, &find_first_cast_succ_block);
}
```

`malloc` and `calloc` are from `*c_void`s, so I turned that code into the current code.

It seems like the destination should be non-`*c_void` here, though.  Yet when I try to flip them, it crashes in `CVoidPtr::checked` where it checks if it actually is a `*c_void`.

I think this `destination` is the destination of the `malloc`, etc. call, so what it returns, `*c_void`, not the destination of the cast.  In the flipped direction, the first arg of `free`, etc. is the `*c_void`.

Reply:
> I think this destination is the destination of the malloc, etc. call, so what it returns, *c_void, not the destination of the cast. In the flipped direction, the first arg of free, etc. is the *c_void.

Right, I conveniently forgot this